types/part_set.go

```diff
@@ -534,22 +548,87 @@ func (ps *PartSet) AddPartWithoutProof(part *Part) (bool, error) {
 	}
 
 	// If part already exists, return false.
-	if ps.parts[part.Index] != nil {
+	if ps.partsBitArray.GetIndex(int(part.Index)) {
 		return false, nil
 	}
 
-	// Add part
-	ps.parts[part.Index] = part
+	// Calculate buffer position and copy part data
+	start := int(part.Index) * ps.partSize
+	end := start + len(part.Bytes)
+
+	// Ensure we don't exceed buffer bounds
+	if end > len(ps.buffer) {
+		return false, fmt.Errorf("part data exceeds buffer bounds")
+	}
+
+	copy(ps.buffer[start:end], part.Bytes)
```

Comment:
can we add a test for this that tests multiple part sizes and see if it completely works. This is the trickiest line of the PR I feel